sql/src/main/java/io/crate/execution/ddl/tables/AlterTableOperation.java

```diff
@@ -236,6 +284,211 @@ private Session session() {
         return result;
     }
 
+    private CompletableFuture<Long> executeAlterTableChangeNumberOfShards(AlterTableAnalyzedStatement analysis) {
+        final DocTableInfo table = analysis.table();
+        final boolean isPartitioned = table.isPartitioned();
+        String sourceIndexName, sourceIndexAlias;
+        if (isPartitioned) {
+            Optional<PartitionName> partitionName = analysis.partitionName();
+            assert partitionName.isPresent() : "Resizing operations for partitioned tables " +
+                                               "are only supported at partition level";
+            sourceIndexName = partitionName.get().asIndexName();
+            sourceIndexAlias = table.ident().indexName();
+        } else {
+            sourceIndexName = table.ident().indexName();
+            sourceIndexAlias = null;
+        }
+        String targetIndexName = SHRINK_PREFIX + sourceIndexName;
+
+        int targetNumberOfShards = analysis.tableParameter().settings().getAsInt(SETTING_NUMBER_OF_SHARDS, 0);
+        validateForResize(sourceIndexName, targetNumberOfShards);
+
+        List<ChainableAction<Long>> actions = new ArrayList<>(7);
+
+        // pickup node
+        DiscoveryNode resizeNode = getNodeForResize();
+
+        // change allocation
+        actions.add(new ChainableAction<>(
+            () -> restrictAllocationToNode(resizeNode.getName(), sourceIndexName),
+            () -> revertAllocation(sourceIndexName)
+        ));
+
+        // shrink index
+        actions.add(new ChainableAction<>(
+            () -> shrinkIndex(sourceIndexName, sourceIndexAlias, targetIndexName, targetNumberOfShards),
+            () -> deleteIndex(targetIndexName)
+        ));
+
+        // close index
+        actions.add(new ChainableAction<>(
+            () -> closeTable(sourceIndexName, targetIndexName),
+            () -> openTable(sourceIndexName, targetIndexName)
+        ));
+
+        // index rename
+        final String backupIndexName = ".backup" + sourceIndexName;
```

Comment:
also here, I'd prefer to use a `.` human separator at the prefix end, also prefix could be a static var..